File: java/src/main/java/org/example/algorithms/sliding_window/LongestSubstringWithoutRepeatingCharacters.java
package org.example.algorithms.sliding_window;

public class LongestSubstringWithoutRepeatingCharacters {
    // https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
    public static int lengthOfLongestSubstring(String s) {
        int left = 0;
        int max = 0;
        int[] frequency = new int[128];

        for (int right = 0; right < s.length(); right++) {
            frequency[s.charAt(right)]++;

            while(frequency[s.charAt(right)] > 1) {
                frequency[s.charAt(left)]--;
                left++;
            }

            max = Math.max(max, right - left + 1);
        }

        return max;
    }
}

File: go/algorithms/sliding_window/longest_substring_without_repeating_characters.go
package slidingwindow

func lengthOfLongestSubstring(s string) int {
	left := 0
	result := 0
	frequency := make(map[byte]int)

	for right := 0; right < len(s); right++ {
		if value, ok := frequency[s[right]]; ok && value >= left {
			left = value + 1
		}

		result = max(result, right-left+1)
		frequency[s[right]] = right
	}

	return result
}
